### PR TITLE
Add Opc diff unit test

### PR DESCRIPTION
**Pull request by Icermli**
Unit Tests for OpcDiff

### CODE REVIEW

File: src/test/scala/vsys/blockchain/state/opcdiffs/AssertOpcDiffTest.scala [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+      Left(GenericError(s"Invalid Assert (eq): DataEntry " +
+        s"${DataEntry(Longs.toByteArray(1), DataType.Amount).json} " +
+        s"is not equal to ${DataEntry(Ints.toByteArray(1), DataType.Int32).json}")))
+
+    val tx = RegisterContractTransaction.create(PrivateKeyAccount(Array.fill(TransactionParser.KeyLength)(0)),
+      ContractPermitted.contract, Seq(DataEntry(Longs.toByteArray(-1), DataType.Amount)),
+      "vsys", 10000L, 100, 1L)
+
+    AssertOpcDiff.isCallerOrigin(ExecutionContext.fromRegConTx(state, TestFunctionalitySettings.Enabled, Option(0L),
+      1L, 1, tx.right.get).right.get)(
```
Comment (ncying) [inline src/test/scala/vsys/blockchain/state/opcdiffs/AssertOpcDiffTest.scala:69]: all `ExecutionContext.fromRegConTx(state, TestFunctionalitySettings.Enabled, Option(0L), 1L, 1, tx.right.get).right.get` can be picked out as value context to simplify the code

File: src/test/scala/vsys/blockchain/state/opcdiffs/AssertOpcDiffTest.scala [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+      DataEntry(PrivateKeyAccount(Array.fill(TransactionParser.KeyLength)(0)).toAddress.bytes.arr,
```
Comment (ncying) [inline src/test/scala/vsys/blockchain/state/opcdiffs/AssertOpcDiffTest.scala:70]: `PrivateKeyAccount(Array.fill(TransactionParser.KeyLength)(0))` similar, key transformation will waste time if do it repeatly.

File: src/test/scala/vsys/blockchain/state/opcdiffs/BasicOpcDiffTest.scala [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+    BasicOpcDiff.concat(
+      DataEntry(Array.fill(maxShortBytesLength){0}, DataType.ShortBytes),
+      DataEntry(Array.fill(1){0}, DataType.ShortBytes),
+      Seq.empty, 0) should be (Left(InvalidDataEntry))
+
+    BasicOpcDiff.constantGet(Array(DataType.Amount.id.toByte) ++ Longs.toByteArray(1),
+      Seq.empty, 0) should be (Right(Seq(DataEntry(Longs.toByteArray(1), DataType.Amount))))
+    BasicOpcDiff.constantGet(Array(DataType.Amount.id.toByte) ++ Ints.toByteArray(1),
+      Seq.empty, 0) should be (Left(InvalidDataEntry))
+  }
```
Comment (ncying) [inline src/test/scala/vsys/blockchain/state/opcdiffs/BasicOpcDiffTest.scala:125]: We need to add `pointer` in the original code and test the usage here.

File: src/test/scala/vsys/blockchain/state/opcdiffs/CompareOpcDiffTest.scala [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+import com.google.common.primitives.{Ints, Longs}
+import org.scalatest.prop.{GeneratorDrivenPropertyChecks, PropertyChecks}
+import org.scalatest.{Matchers, PropSpec}
+import vsys.blockchain.contract.{DataEntry, DataType}
+import vsys.blockchain.transaction.TransactionGen
+import vsys.blockchain.transaction.ValidationError.{ContractDataTypeMismatch, ContractUnsupportedOPC}
+
+import scala.util.{Left, Right}
+
+class CompareOpcDiffTest extends PropSpec with PropertyChecks with GeneratorDrivenPropertyChecks with Matchers with TransactionGen {
```
Comment (ncying) [inline src/test/scala/vsys/blockchain/state/opcdiffs/CompareOpcDiffTest.scala:12]: `TransactionGen`?

File: src/test/scala/vsys/blockchain/state/opcdiffs/ReturnOpcDiffTest.scala [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+import vsys.blockchain.contract.{ContractPermitted, DataEntry, DataType, ExecutionContext}
+import vsys.blockchain.state.diffs.newState
+import vsys.blockchain.transaction.{TransactionGen, TransactionParser}
+import vsys.blockchain.transaction.ValidationError.ContractUnsupportedOPC
+import vsys.blockchain.transaction.contract.RegisterContractTransaction
+import vsys.settings.TestFunctionalitySettings
+
+import scala.util.Left
+
+class ReturnOpcDiffTest extends PropSpec with PropertyChecks with GeneratorDrivenPropertyChecks with Matchers with TransactionGen {
```
Comment (ncying) [inline src/test/scala/vsys/blockchain/state/opcdiffs/ReturnOpcDiffTest.scala:16]: we only use create fun here,  all `TransactionGen` should be useless?

File: src/test/scala/vsys/blockchain/state/opcdiffs/ReturnOpcDiffTest.scala [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+
+  property("test return opcs") {
+
+    val state = newState()
+
+    val tx = RegisterContractTransaction.create(PrivateKeyAccount(Array.fill(TransactionParser.KeyLength)(0)),
+      ContractPermitted.contract, Seq(DataEntry(Longs.toByteArray(-1), DataType.Amount)),
+      "vsys", 10000L, 100, 1L)
```
Comment (ncying) [inline src/test/scala/vsys/blockchain/state/opcdiffs/ReturnOpcDiffTest.scala:24]: some commonly used code can be picked out as a file.

File: src/test/scala/vsys/blockchain/state/opcdiffs/BasicOpcDiffTest.scala (hunk 3 above)
Comment (Icermli) [reply src/test/scala/vsys/blockchain/state/opcdiffs/BasicOpcDiffTest.scala:126]: `pointer` in the original code? What is that?

Comment (ncying) [reply src/test/scala/vsys/blockchain/state/opcdiffs/BasicOpcDiffTest.scala:126]: in `BasicOpcDiff.scala`, we may check the pointer range, then test the incorrect pointer here.

Comment (Icermli) [reply src/test/scala/vsys/blockchain/state/opcdiffs/BasicOpcDiffTest.scala:126]: yea, I checked when the pointer is out of range

**Review by ncying — APPROVED** [review]
LGTM

**Review by faddat — COMMENTED** [review]
Comments in the code would make reviews easier and improve readability.

**Pull request merged**